linker.go

```diff
@@ -1,57 +1,149 @@
+// so when an instruction in one ELF program section wants to jump to
+// a function in another, the linker needs to pull in the bytecode
+// (and BTF info) of the target function and concatenate the instruction
+// streams together.
 //
-// Each library may contain multiple functions / labels, and is only linked
-// if prog references one of these functions.
+// Later on in the pipeline, all call sites are fixed up with relative jumps
+// within this newly-created instruction stream to then finally hand off to
+// the kernel with BPF_PROG_LOAD.
 //
-// Libraries also linked.
-func link(prog *ProgramSpec, libs []*ProgramSpec) error {
-	var (
-		linked  = make(map[*ProgramSpec]bool)
-		pending = []asm.Instructions{prog.Instructions}
-		insns   asm.Instructions
-	)
-	for len(pending) > 0 {
-		insns, pending = pending[0], pending[1:]
-		for _, lib := range libs {
-			if linked[lib] {
+// Each function is denoted by an ELF symbol and the compiler takes care of
+// register setup before each jump instruction.
+
+// findNeighbours finds bpf-to-bpf calls in all given progs.
+// It populates each ProgramSpec's neighbour list with pointers
+// to all other programs referenced by its bytecode.
+func findNeighbours(progs map[string]*ProgramSpec) error {
+	// Check all ProgramSpecs in the collection against each other.
+	for _, prog := range progs {
+		for _, dep := range progs {
+			// Don't link a program against itself.
+			if prog == dep {
 				continue
 			}
 
-			needed, err := needSection(insns, lib.Instructions)
+			need, err := needProg(prog.Instructions, dep.Instructions)
 			if err != nil {
-				return fmt.Errorf("linking %s: %w", lib.Name, err)
+				return fmt.Errorf("dependency-checking program '%s' and '%s': %w", prog.Name, dep.Name, err)
 			}
 
-			if !needed {
+			if !need {
 				continue
 			}
 
-			linked[lib] = true
-			prog.Instructions = append(prog.Instructions, lib.Instructions...)
-			pending = append(pending, lib.Instructions)
-
-			if prog.BTF != nil && lib.BTF != nil {
-				if err := prog.BTF.Append(lib.BTF); err != nil {
-					return fmt.Errorf("linking BTF of %s: %w", lib.Name, err)
-				}
-			}
+			// Register a direct dependency.
+			prog.neighbours = append(prog.neighbours, dep)
 		}
 	}
 
 	return nil
 }
 
-func needSection(insns, section asm.Instructions) (bool, error) {
-	// A map of symbols to the libraries which contain them.
-	symbols, err := section.SymbolOffsets()
+// collectInstructions returns the instruction stream of all progs in order.
+func collectInstructions(progs []*ProgramSpec) asm.Instructions {
+	var out asm.Instructions
+
+	for _, prog := range progs {
+		out = append(out, prog.Instructions...)
+	}
+
+	return out
+}
+
+// collectFuncInfos returns the BTF func infos of all progs in order.
+func collectFuncInfos(progs []*ProgramSpec) ([]byte, error) {
+	if len(progs) == 0 {
+		return nil, nil
+	}
+
+	var off uint64
+
+	buf := bytes.NewBuffer(make([]byte, 0, binary.Size(&btf.FuncInfo{})*len(progs)))
```

Comment:
We'll get rid of `FuncInfo(s)` when eliminating `btf.Program` anyway. I considered this during the initial implementation, and just tried another implementation of `FuncInfos.Marshal()`, but it's basically the same code iterating over `infos` instead of `progs`. The callsite here doesn't become much cleaner since the function's offset within the instruction stream still needs to be tracked and populated; it's just double the amount of code to accomplish the same thing. I'll leave this be for now.